records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -136,6 +136,52 @@ Keeping track of emails sent per service during the `[bounce-rate-window]` shoul
 #### Storing the number of hard bounces
 Hard bounces can be detected as part of the `process-ses-result` celery task, and can be stored in redis for the `[bounce-rate-window]`.
 
+## Redis Implementation Details
+--------------------------------
+
+### Keys in Redis
+
+In order to not hit the database directly we are going to store two keys for each service. These two keys are stored as a Redis sorted set. The `{key:value}` pair input will be `{epoch: epoch}`
+
+`total_notifications:{service_id}`\
+`hard_bounces:{service_id}`
+
+$$\text{Bounce Rate} = \frac{\text{length}( \text{hard bounces:\{service id\}} )}{ \text{length}( \text{total notifications:\{service id\}} )}$$
+
+
+The reason we put a `{k:v}` as epoch is because the `k` must be unique. We add to total_notifications every single time we send a notification. We keep the key as `epoch` so we might calculate the bounce rate using a sliding window. We can find the length of the above two keys, where only the keys in the last 24 hours (this is adjustable) are included in the calculation.
+
+### Seeding Redis
+
+#### Seeding the keys:
+If we have to seed `total_notifications` or `hard_bounces` per service_id, we would have to take every single notification added in the last X-hours (X being 24 etc) and add to the total_notification_service_id. This means looping through every single notification row for the last 24 hours and adding it to the list.
+
+Instead we will find the total_notification and hard_bounces for each of the last 24 hours and add that to redis. For each `service_id` we will now have 50 keys - 48 keys, one for total_notifications / hour and one of hard_bounces/ hour.
+
+The following keys:\
+`total_notifications_service_id_{24_hours}`\
+`total_notifications_service_id_{23_hours}`\
+....\
+`hard_bounces_service_id_{1_hour}`\
+....\
+`total_notifications:{service_id}`\
+`hard_bounces:{service_id}`
+
+The 48 keys would expire within the 24 hour mark (or whatever we chose the sliding window to be)
```

Comment:
So as time ticks on, the "hour" keys will get dropped from the calculation each hour - very cool!